lib/staccato/adapter/batch.rb

```diff
@@ -0,0 +1,80 @@
+require 'thread'
+
+module Staccato
+  module Adapter
+    class Batch
+      BatchTooBigError = Class.new(StandardError)
+
+      DEFAULT_MAX_SIZE = 20
+      FLUSHING_TIMEOUT = 10
+
+      attr_reader :adapter
+
+      def initialize(adapter=Staccato.default_adapter, size: DEFAULT_MAX_SIZE, flush_timeout: FLUSHING_TIMEOUT)
+        if size > DEFAULT_MAX_SIZE
+          raise BatchTooBigError, "Batch size #{size} is larger than Google accepts."
+        end
+
+        @size = size
+        @adapter = adapter.new(Staccato.ga_batch_uri)
+
+        @queue = SizedQueue.new(@size)
+        @flushing_thread = Thread.new { loop { perform_flush; sleep(flush_timeout) } }
```

Comment:
@bots-business I could change this PR so that no `flushing_thread` is created if the `flush_timeout` is explicitly set to `nil`. This way the tracker and adapter are still thread-safe, but no additional thread is created to perform auto-flushing of the queue.